metrics.go

```diff
@@ -0,0 +1,112 @@
+
+import (
+	"fmt"
+	"net/http"
+
+	"github.com/kansaslabs/x/out"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+)
+
+var (
+	pmWorkers     prometheus.Gauge // number of available workers
+	pmQueueSize   prometheus.Gauge // number of tasks in the queue awaiting handling
+	pmPercentFull prometheus.Gauge // the percent of the queue that is full * 100
+	// pmPercentSuccess *prometheus.GaugeVec     // the percent of tasks successfully completed, labeled by task
+	pmTasksSucceeded *prometheus.CounterVec   // the count of successfully completed tasks, labeled by task type
+	pmTasksFailed    *prometheus.CounterVec   // the count of failed tasks, labeled by task type
+	pmTaskLatency    *prometheus.HistogramVec // the time it is taking for tasks to complete, labeled by task type, success, and failure
+)
+
+const (
+	pmNamespace = "radish"
+	pmSubsystem = "queue"
+)
+
+func initMetrics() {
+	pmWorkers = prometheus.NewGauge(prometheus.GaugeOpts{
+		Namespace: pmNamespace,
+		Subsystem: pmSubsystem,
+		Name:      "workers",
+		Help:      "The number of available workers",
+	})
+
+	pmQueueSize = prometheus.NewGauge(prometheus.GaugeOpts{
+		Namespace: pmNamespace,
+		Subsystem: pmSubsystem,
+		Name:      "queue_size",
+		Help:      "number of tasks in the queue awaiting handling",
+	})
+
+	pmPercentFull = prometheus.NewGauge(prometheus.GaugeOpts{
+		Namespace: pmNamespace,
+		Subsystem: pmSubsystem,
+		Name:      "percent_full",
+		Help:      "the percent of the queue that is already full",
+	})
+
+	// TODO: Come back to this; would need to keep track of global tasks?
+	// pmPercentSuccess = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+	// 	Namespace: pmNamespace,
+	// 	Subsystem: pmSubsystem,
+	// 	Name:      "percent_success",
+	// 	Help:      "the percent of tasks successfully completed, labeled by task",
+	// }, []string{"task"})
+
+	pmTasksSucceeded = prometheus.NewCounterVec(prometheus.CounterOpts{
+		Namespace: pmNamespace,
+		Subsystem: pmSubsystem,
+		Name:      "tasks_succeeded",
+		Help:      "the count of tasks successfully completed, labeled by task type",
+	}, []string{"task"})
+
+	pmTasksFailed = prometheus.NewCounterVec(prometheus.CounterOpts{
+		Namespace: pmNamespace,
+		Subsystem: pmSubsystem,
+		Name:      "tasks_failed",
+		Help:      "the count of failed tasks, labeled by task type",
+	}, []string{"task"})
+
+	pmTaskLatency = prometheus.NewHistogramVec(prometheus.HistogramOpts{
+		Namespace: pmNamespace,
+		Subsystem: pmSubsystem,
+		Name:      "task_latency",
+		Help:      "time to task completion, labeled by task type, success, and failure",
+	}, []string{"task", "result"})
+}
+
+func serveMetrics(metricsAddr string) {
+	out.Status("serving prometheus metrics at http://%s/metrics", metricsAddr)
+	http.Handle("/metrics", promhttp.Handler())
+	if err := http.ListenAndServe(metricsAddr, nil); err != nil {
+		out.Warne(err)
+	}
+}
+
+func registerMetrics() error {
```

Comment:
The only remaining question is when do we call this internal function -- or do we make it public? We don't want to use `prometheus.MustRegister` -- which will panic if the user has a different version of Prometheus than Radish, so we want the ability to return an error at runtime.